### DISCUSSION TITLE
What's the idea behind the unusual versioning scheme?

### DISCUSSION THREAD

**Discussion by benjamin-hempel** [AdamWhiteHat/BigDecimal#65] [Questions or Help] [answered]
Hello,

I was wondering what the idea behind the rather unusual versioning scheme of this library is. Why doesn't it use semantic versioning, like the vast majority of other libraries out there?

Thanks in advance!

Kind regards
Benjamin

**Comment by AdamWhiteHat** [comment, marked as answer]
Well, it does follow semantic versioning... now.

<details>
<summary>But originally, it didn't, and the version reflected the date that the assembly was built on.</summary>

<p>

 &nbsp;&nbsp; :information_source: &nbsp;&nbsp; Specifically, the format was: `yyyy.DOY.HHmm`, where DOY was the day-of-the-year, a value in the range of 1 to 365. 

</p>

</details>

Obviously, this was suboptimal; a lot of people were using my library, and I needed to be able to communicate a change that might break implementations that relied upon its behavior, from ones that didn't.

When I decided to adopt semantic versioning. I just ripped out the mechanism that automatically changed the version number on each build and began manually updating the version numbers as needed for each release, in line with semantic versioning.

At that point I had been releasing the library on NuGet for several years.
In order to ensure that NuGet continued to recognize my new builds as the latest version, and that it displayed my builds in the correct order, I could not 'reset' my version number back to version 2.0.0 or whatever and begin incrementing from there like I intended to do; I had to publish builds with version number(s) that were strictly increasing. 
NuGet orders your packages solely on the version number, regardless of when you uploaded them.

So the major version number remained at 2023 for some time, but I did bump the minor version number to 1000 to signify the change--a number that by the previous scheme could not have been reached, as the minor version number reflected the day of the year it was build on.

And so the version numbers were of the form 2023.1000.* for some time. Upon the next breaking change I bumped the major version number to 2025--a year that had not yet come to pass--to try and signify that it no longer reflects the date. 

But it still LOOKED like a date, and 2025 would soon become the current year, so I took advantage of the next breaking change to bump the  major version number all the way up to  3000, and rolling over the minor and patch version numbers back to zero. 

Hopefully 3000 is sufficiently far enough removed to break any perceived association with dates that people might have.

And its likely to remain at 3000.*.* for some time; I wont be releasing any more breaking changes if it can be helped.

**Comment by benjamin-hempel** [comment]
Hi Adam,

thank you for the comprehensive explanation. We just noticed it when adding the package as a dependency of [NCalc](https://github.com/ncalc/ncalc), but it does of course make sense to ease the scheme back into semantic versioning so that NuGet still recognizes new package versions as such.

Thanks again and happy holidays!

Kind regards
Benjamin
